mailpoet/lib/Automation/Engine/Builder/DuplicateWorkflowController.php

```diff
@@ -0,0 +1,91 @@
+<?php declare(strict_types = 1);
+
+namespace MailPoet\Automation\Engine\Builder;
+
+use MailPoet\Automation\Engine\Data\NextStep;
+use MailPoet\Automation\Engine\Data\Step;
+use MailPoet\Automation\Engine\Data\Workflow;
+use MailPoet\Automation\Engine\Exceptions;
+use MailPoet\Automation\Engine\Exceptions\InvalidStateException;
+use MailPoet\Automation\Engine\Storage\WorkflowStorage;
+use MailPoet\Automation\Engine\WordPress;
+use MailPoet\Util\Security;
+
+class DuplicateWorkflowController {
+  /** @var WordPress */
+  private $wordPress;
+
+  /** @var WorkflowStorage */
+  private $workflowStorage;
+
+  public function __construct(
+    WordPress $wordPress,
+    WorkflowStorage $workflowStorage
+  ) {
+    $this->wordPress = $wordPress;
+    $this->workflowStorage = $workflowStorage;
+  }
+
+  public function duplicateWorkflow(int $id): Workflow {
+    $workflow = $this->workflowStorage->getWorkflow($id);
+    if (!$workflow) {
+      throw Exceptions::workflowNotFound($id);
+    }
+
+    $duplicate = new Workflow(
+      $this->getName($workflow->getName()),
+      $this->getSteps($workflow->getSteps()),
+      $this->wordPress->wpGetCurrentUser()
+    );
+    $duplicate->setStatus(Workflow::STATUS_DRAFT);
+
+    $workflowId = $this->workflowStorage->createWorkflow($duplicate);
+    $savedWorkflow = $this->workflowStorage->getWorkflow($workflowId);
+    if (!$savedWorkflow) {
+      throw new InvalidStateException('Workflow not found.');
+    }
+    return $savedWorkflow;
+  }
+
+  private function getName(string $name): string {
+    // translators: %s is the original workflow name.
+    $newName = sprintf(__('Copy of %s', 'mailpoet'), $name);
+    $maxLength = $this->workflowStorage->getNameColumnLength();
+    if (strlen($newName) > $maxLength) {
```

Comment:
Here, if we were talking Unicode, we should actually use `mb_strlen`. But I don't know for sure... likely, for that to be completely correct, the users would need to use `utf8mb4` DB encoding, but we have no control over that.

So still thinking if 255 chars or potentially bytes, is OK, or if we should increase the column size 🤔